### PR TITLE
[FIXED JENKINS-40793] Add location information to validation errors

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-40793](https://issues.jenkins-ci.org/browse/JENKINS-40793)

Specifically to the JSON output. This also moves us to not doing
conversion between JSONObject and JsonNode, which is nice. There's
probably some cleanup/optimization that could still be done in
JSONParser.

cc @reviewbybees esp @kzantow @rsandell

### CODE REVIEW

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/Converter.groovy [hunk 1]
```diff
@@ -64,8 +67,10 @@ public class Converter {
...
      */
     public static ProcessingReport validateJSONAgainstSchema(JSONObject origJson) throws ProcessingException {
-        JsonNode jsonNode = jacksonJSONFromJSONObject(origJson);
+        return validateJSONAgainstSchema(jacksonJSONFromJSONObject(origJson))
+    }
 
+    public static ProcessingReport validateJSONAgainstSchema(JsonNode jsonNode) throws ProcessingException {
         JsonSchema schema = ASTSchema.getJSONSchema();
 
         return schema.validate(jsonNode)
```
Comment (kzantow) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/Converter.groovy]: although not really part of this PR, I've noticed some validation errors while converting from groovy are caught that are not when converting from json, could this be validated the same way here? One example is duplicate parallel names are allowed from json -> pipeline but not pipeline -> json

Comment (abayer) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/Converter.groovy]: Huh, interesting. Lemme poke at that.

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/JSONErrorCollector.groovy [hunk 2]
```diff
@@ -49,6 +57,38 @@ public class JSONErrorCollector extends ErrorCollector {
...
+
+        public JSONObject getJsonError() {
+            JSONObject o = new JSONObject()
+            JSONArray a = new JSONArray()
+
+            location.iterator().each { t ->
+                a.add(t.toString())
+            }
+
+            o.accumulate("location", a)
```
Comment (kzantow) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/JSONErrorCollector.groovy]: ehh? `put` instead?

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/SourceUnitErrorCollector.groovy [hunk 3]
```diff
@@ -45,6 +47,19 @@ public class SourceUnitErrorCollector extends ErrorCollector {
     }
 
     @Override
+    public JSONArray asJson() {
+        JSONArray a = new JSONArray()
+
+        errorsAsStrings().each {
+            JSONObject o = new JSONObject()
+            o.accumulate("error", it)
```
Comment (kzantow) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/SourceUnitErrorCollector.groovy]: `put` would be more clear?

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/endpoints/ModelConverterAction.java [hunk 4]
```diff
@@ -23,15 +23,16 @@
  */
 package org.jenkinsci.plugins.pipeline.modeldefinition.endpoints;
 
+import com.fasterxml.jackson.databind.JsonNode;
+import com.github.fge.jsonschema.tree.SimpleJsonTree;
+import com.github.fge.jsonschema.util.JsonLoader;
+import com.google.common.collect.ImmutableList;
```
Comment (kzantow) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/endpoints/ModelConverterAction.java]: nooooo!

**Comment by kzantow** [comment]
Argh FindBugs issues...

**Comment by rsandell** [comment]
:bee:

**Comment by rsandell** [comment]
I can't undestand half of what this new Json api does, but a half blind :bee: anyway ;)

**Comment by abayer** [comment]
...fair enough. =)
